File: mcomix/file_handler.py
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

from pathlib import Path

from loguru import logger

from mcomix.archive_extractor import Extractor
from mcomix.enums import FileSortDirection, FileSortType, FileTypes
from mcomix.formats.archive import ArchiveSupported
from mcomix.file_provider import GetFileProvider
from mcomix.image_files import ImageFiles
from mcomix.image_handler import ImageHandler
from mcomix.lib.events import Events, EventType
from mcomix.lib.metaclass import SingleInstanceMetaClass
from mcomix.preferences import config
from mcomix.sort.sort_alphanumeric import SortAlphanumeric

from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from mcomix.main_window import MainWindow


class FileHandler(metaclass=SingleInstanceMetaClass):
    """
    The FileHandler keeps track of the actual files/archives opened.
    While ImageHandler takes care of pages/images, this class provides
    the raw file names for archive members and image files, extracts
    archives, and lists directories for image files
    """

    def __init__(self, window: MainWindow):
        super().__init__()

        self.__events = Events()
        self.__events.add_event(EventType.FILE_EXTRACTED, self._extracted_file)
        self.__events.add_event(EventType.FILE_LISTED, self._file_listed)
        self.__events.add_event(EventType.KB_FILE_CLOSE, self.close_file)
        self.__events.add_event(EventType.KB_FILE_REFRESH, self.refresh_file)
        self.__events.add_event(EventType.KB_OPEN_ARCHIVE_DIRECTION, self.open_archive_direction)

        self.__image_handler = ImageHandler()
        self.__image_files = ImageFiles()

        #: Indicates if files/archives are currently loaded/loading.
        self.__file_loaded = False
        self.__file_loading = False
        #: False if current file is not an archive, or unrecognized format.
        self.__is_archive = False

        #: Either path to the current archive, or first file in image list.
        #: This is B{not} the path to the currently open page.
        self.__current_file = None
        #: Reference to L{MainWindow}.
        self.__window = window
        #: Path to opened archive file, or directory containing current images.
        self.__base_path = None
        #: Archive extractor.
        self.__extractor = None
        #: Provides a list of available files/archives in the open directory.
        self.__file_provider_chooser = GetFileProvider()
        self.__file_provider = None

        self.__start_page = 1

        self.__open_first_page = None

        self.update_opening_behavior()

    def update_opening_behavior(self):
        self.__open_first_page = 1 if config['OPEN_FIRST_PAGE'] else -1

    def refresh_file(self):
        """
        Closes the current file(s)/archive and reloads them
        """

        if not self.__file_loaded:
            return

        current_file = self.get_real_path()
        if self.__is_archive:
            start_page = self.__image_handler.get_current_page()
        else:
            start_page = 1
        self.open_file(current_file, start_page)

    def open_file_init(self, paths: list, start_page: int = 1):
        """
        Open the first file pointed to in <paths>,
        and initialize the file provider.
        """

        if not paths:
            return

        self._initialize_fileprovider(paths=paths)
        self.open_file(paths[0], start_page)

    def open_file(self, path: Path, start_page: int = 1):
        """
        Open the file pointed to by <path>.
        If <start_page> is not set we set the current
        page to 1 (first page), if it is set we set the current page to the
        value of <start_page>. If <start_page> is non-positive it means the
        last image.
        """

        self._close()

        self.__is_archive = ArchiveSupported.is_archive_file(path)
        self.__start_page = start_page
        self.__current_file = path

        # Actually open the file(s)/archive passed in path.
        if self.__is_archive:
            self.__base_path = path
            self._open_archive(self.__current_file)
            self.__file_loading = True
        else:
            image_files = self.__file_provider.list_files(mode=FileTypes.IMAGES)
            if self.__current_file.is_dir():
                self.__base_path = self.__current_file
            else:
                self.__base_path = self.__current_file.parent
            self._archive_opened(image_files)

    def _archive_opened(self, image_files: list):
        """
        Called once the archive has been opened and its contents listed
        """

        self.__image_files.set_image_files(image_files)
        self.file_opened()

        if not image_files:
            logger.error(f'No images in "{self.__current_file.name}"')
            return

        if self.__is_archive:
            self.__extractor.extract()

            start_page = self.__start_page
        else:
            # If no extraction is required, mark all files as available.
            for img in image_files:
                self.__image_handler.file_available(img)

            if self.__current_file.is_dir():
                # if the current_file is a directory then start at the first image
                start_page = 1
            else:
                # Set start_page to the same as current_file.
                start_page = self.__image_files.get_page_from_path(self.__current_file)

        self.__window.set_page(start_page)

    def file_opened(self):
        """
        Called when a new set of files has successfully been opened
        """

        self.__file_loaded = True

        self.__events.run_events(EventType.FILE_OPENED)

    def file_closed(self):
        """
        Called when the current file has been closed
        """

        self.__events.run_events(EventType.FILE_CLOSED)

    def close_file(self):
        """
        Close the currently opened file and its provider
        """

        self._close(close_provider=True)

    def _close(self, close_provider: bool = False):
        """
        Run tasks for "closing" the currently opened file(s)
        """

        if self.__file_loaded or self.__file_loading:
            if close_provider:
                self.__file_provider = None
            if self.__is_archive:
                self.__extractor.close()
            self.__image_handler.cleanup()
            self.__file_loaded = False
            self.__file_loading = False
            self.__is_archive = False
            self.__current_file = None
            self.__base_path = None
            self.file_closed()

    def _initialize_fileprovider(self, paths: list):
        """
        Initialize the file provider based on the total number of file paths
        in <paths>. If <paths> is a list with more than one entry, assumes that
        only the files in the list should be available. If <paths> is a list with
        a single path, assume that it is either a directory or an image file,
        and all files in that directory should be opened.

        :param path: List of file names, or single file/directory as string.
        """

        self.__file_provider = self.__file_provider_chooser.get_file_provider(paths)

    def _open_archive(self, path: Path):
        """
        Extracts all images in the archive passed in path
        """

        try:
            self.__extractor = Extractor(path)
            self.__extractor.list_contents()
        except Exception as ex:
            logger.error(f'failed to open archive: {path}')
            logger.debug(f'Exception: {ex}')
            raise

    def _file_listed(self, files: list):
        if not self.__file_loading:
            return
        self.__file_loading = False

        self._sort_archive_images(files)
        self._archive_opened(files)

    def _sort_archive_images(self, filelist: list):
        """
        Sorts the image list passed in C{filelist} based on the sorting preference option
        """

        # sort files
        match config['SORT_ARCHIVE_BY']:
            case FileSortType.NAME.value:
                SortAlphanumeric(filelist)
            case FileSortType.NAME_LITERAL.value:
                filelist.sort()

        # sort files order
        if config['SORT_ARCHIVE_ORDER'] == FileSortDirection.DESCENDING.value:
            filelist.reverse()

    def is_file_loaded(self) -> bool:
        return self.__file_loaded

    def is_archive(self):
        """
        Returns True if the current opened file an archive.
        """

        return self.__is_archive

    def get_base_path(self):
        """
        Returns the path to opened archive file, or directory containing current images.
        """

        return self.__base_path

    def _get_file_list(self):
        """
        Get a list of all archives in the same diretory as the opened archive
        """

        return self.__file_provider.list_files(mode=FileTypes.ARCHIVES)

    def get_file_number(self):
        """
        Get the current position of the opened archive and the total archives in the
        current directory. If the current opened file is not an archive do nothing.

        :returns a tuple of (current file number, total files)
        """

        if not self.__is_archive:
            # No file numbers for images.
            return 0, 0

        file_list = self._get_file_list()
        current_index = file_list.index(self.__current_file)

        return current_index + 1, len(file_list)

    def get_real_path(self):
        """
        Return the "real" path to the currently viewed file, i.e. the
        full path to the archive or the full path to the currently viewed image
        """

        if self.__is_archive:
            return self.__base_path

        return self.__image_handler.get_path_to_page()

    def open_archive_direction(self, forward: bool):
        """
        Opens the archive that comes directly after the currently loaded
        archive in that archive's directory listing if forward=True else
        opens the archive that comes directly before the currently loaded
        archive in that archive's directory listing. sorted alphabetically.

        :returns True if a new archive was opened, False otherwise
        """

        if not self.__is_archive:
            return False

        files = self._get_file_list()
        if self.__base_path not in files:
            return False

        current_index = files.index(self.__base_path)

        if forward:
            next_file = files[current_index + 1:]
            next_page = 1
        else:
            next_file = reversed(files[:current_index])
            next_page = self.__open_first_page

        for path in next_file:
            self._close()
            self.open_file(path, next_page)
            return True

        return False

    def _extracted_file(self, filename: Path):
        """
        Called when the extractor finishes extracting the file at
        <name>. This name is relative to the temporary directory
        the files were extracted to
        """

        if not self.__file_loaded:
            return
        self.__events.run_events(EventType.FILE_AVAILABLE, {'filename': filename})

File: src/vfs/file-handler.cxx
/**
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <algorithm>
#include <filesystem>
#include <memory>
#include <span>
#include <vector>

#include <ztd/ztd.hxx>

#include "settings/settings.hxx"

#include "vfs/extractor.hxx"
#include "vfs/file-handler.hxx"
#include "vfs/file-provider.hxx"
#include "vfs/file-supported.hxx"
#include "vfs/image-handler.hxx"

#include "vfs/utils/sort.hxx"

#include "logger.hxx"

vfs::file_handler::file_handler(const std::shared_ptr<config::settings>& settings,
                                const std::shared_ptr<gui::lib::view_state>& view_state) noexcept
    : settings(settings), view_state(view_state)
{
}

void
vfs::file_handler::refresh_opened() noexcept
{
    if (!file_loaded_)
    {
        return;
    }

    page_t start_page;
    const auto current_file = get_real_path();

    if (is_archive_)
    {
        start_page = image_handler_->get_current_page();
    }
    else
    {
        start_page = 1;
    }

    open_file(current_file, start_page);
}

void
vfs::file_handler::open_file_init(const std::span<const std::filesystem::path> filelist,
                                  const page_t start_page) noexcept
{
    if (filelist.empty())
    {
        return;
    }

    initialize_fileprovider(filelist);
    open_file(filelist.front(), start_page);
}

void
vfs::file_handler::open_file(const std::filesystem::path& path, const page_t start_page) noexcept
{
    close();

    image_handler_ = std::make_shared<vfs::image_handler>(settings, view_state);
    image_handler_->signal_page_available().connect([this](const page_t page)
                                                    { signal_page_available().emit(page); });

    is_archive_ = vfs::is_archive(path);
    default_start_page_ = start_page;
    current_file_ = path;

    if (is_archive_)
    {
        base_path_ = path;
        file_loading_ = true;
        open_archive(path);
    }
    else
    {
        const auto image_files = file_provider_->list_files(vfs::file_provider::file_type::images);
        if (std::filesystem::is_directory(current_file_))
        {
            base_path_ = current_file_;
        }
        else
        {
            base_path_ = current_file_.parent_path();
        }
        archive_opened(image_files);
    }
}

void
vfs::file_handler::archive_opened(const std::span<const std::filesystem::path> image_files) noexcept
{
    image_handler_->image_files()->set_image_files(image_files);
    file_opened();

    if (image_files.empty())
    {
        logger::error<logger::vfs>("No images in {}", current_file_.string());
        return;
    }

    page_t start_page = default_start_page_;

    if (is_archive_)
    {
        extractor_->extract();
    }
    else
    {
        // No extraction is required, mark all files as available.
        for (const auto& img : image_files)
        {
            image_handler_->file_available(img);
        }

        if (std::filesystem::is_directory(current_file_))
        {
            // if current_file is a directory then start at the first image
            start_page = 1;
        }
        else
        {
            // Set start_page to the same as current_file
            start_page = image_handler_->image_files()->page_from_path(current_file_);
        }
    }

    signal_page_set().emit(start_page);
}

void
vfs::file_handler::file_opened() noexcept
{
    file_loaded_ = true;

    signal_file_opened().emit();
}

void
vfs::file_handler::file_closed() noexcept
{
    signal_file_closed().emit();
}

void
vfs::file_handler::close_file() noexcept
{
    close(true);
}

std::filesystem::path
vfs::file_handler::current_file() noexcept
{
    return current_file_;
}

void
vfs::file_handler::close(bool close_provider) noexcept
{
    if (file_loaded_ || file_loading_)
    {
        if (close_provider)
        {
            file_provider_ = nullptr;
        }
        if (is_archive_)
        {
            extractor_->close();
        }

        image_handler_ = nullptr;

        file_loaded_ = false;
        file_loading_ = false;
        is_archive_ = false;
        current_file_ = "";
        base_path_ = "";

        file_closed();
    }
}

void
vfs::file_handler::initialize_fileprovider(
    const std::span<const std::filesystem::path> filelist) noexcept
{
    file_provider_ = std::make_unique<file_provider>(filelist);
}

void
vfs::file_handler::open_archive(const std::filesystem::path& archive) noexcept
{
    try
    {
        extractor_ = std::make_unique<vfs::extractor>(archive);

        extractor_->signal_file_extracted().connect([this](const std::filesystem::path& file)
                                                    { extracted_file(file); });
        extractor_->signal_file_listed().connect(
            [this](const std::span<const std::filesystem::path> files) { file_listed(files); });

        extractor_->list();
    }
    catch (const std::exception& ex)
    {
        logger::error<logger::vfs>("failed to open archive: {}", archive.string());
        logger::debug<logger::vfs>("Exception: {}", ex.what());
    }
}

void
vfs::file_handler::file_listed(const std::span<const std::filesystem::path> files) noexcept
{
    if (!file_loading_)
    {
        return;
    }
    file_loading_ = false;

    archive_opened(sort_archive_images(files));
}

std::vector<std::filesystem::path>
vfs::file_handler::sort_archive_images(const std::span<const std::filesystem::path> files) noexcept
{
    auto sorted_files = std::vector<std::filesystem::path>{files.cbegin(), files.cend()};
    vfs::utils::sort_alphanumeric(sorted_files);
    return sorted_files;
}

bool
vfs::file_handler::is_file_loaded() noexcept
{
    return file_loaded_;
}

bool
vfs::file_handler::is_archive() noexcept
{
    return is_archive_;
}

const std::filesystem::path
vfs::file_handler::get_base_path() noexcept
{
    return base_path_;
}

std::span<const std::filesystem::path>
vfs::file_handler::get_file_list() noexcept
{
    return file_provider_->list_files(vfs::file_provider::file_type::archives);
}

const std::array<std::int32_t, 2>
vfs::file_handler::get_file_number() noexcept
{
    if (!is_archive_)
    {
        // No file numbers for images.
        return {0, 0};
    }

    const auto files = get_file_list();
    const auto current_index = current_file_index(files, current_file_);

    if (!current_index)
    {
        return {-1, -1};
    }

    return {static_cast<std::int32_t>(*current_index + 1), static_cast<std::int32_t>(files.size())};
}

const std::filesystem::path
vfs::file_handler::get_real_path() noexcept
{
    if (is_archive_)
    {
        return base_path_;
    }
    return image_handler_->get_path_to_page();
}

bool
vfs::file_handler::open_next_archive() noexcept
{
    if (!is_archive())
    {
        return false;
    }

    const auto files = get_file_list();
    const auto current_index = current_file_index(files, current_file_);
    if (!current_index || (*current_index + 1) == files.size())
    {
        return false;
    }

    const auto next_file = files.at(*current_index + 1);

    open_file(next_file, 1);
    return true;
}

bool
vfs::file_handler::open_prev_archive() noexcept
{
    if (!is_archive())
    {
        return false;
    }

    const auto files = get_file_list();
    const auto current_index = current_file_index(files, current_file_);
    if (!current_index || *current_index == 0)
    {
        return false;
    }

    const auto next_file = files.at(*current_index - 1);

    open_file(next_file, 1);
    return true;
}

bool
vfs::file_handler::open_first_archive() noexcept
{
    if (!is_archive())
    {
        return false;
    }

    const auto files = get_file_list();
    const auto next_file = files.front();

    open_file(next_file, 1);
    return true;
}

bool
vfs::file_handler::open_last_archive() noexcept
{
    if (!is_archive())
    {
        return false;
    }

    const auto files = get_file_list();
    const auto next_file = files.back();

    open_file(next_file, 1);
    return true;
}

void
vfs::file_handler::extracted_file(const std::filesystem::path& filename) noexcept
{
    if (!file_loaded_)
    {
        return;
    }

    image_handler_->file_available(filename);
}
